### DISCUSSION TITLE
SBOM support for Bazel modules

### DISCUSSION THREAD

**Discussion by fmeum** [bazelbuild/bazel#23166] [Ideas]
`rules_license` provides the basic [`license`](https://bazelbuild.github.io/rules_license/latest.html#license) and [`package_info`](https://bazelbuild.github.io/rules_license/latest.html#package_info) rules that serve as the basis for SBOM generation (and other kinds of supply chain analysis) for all kinds of Bazel dependencies. While individual rulesets that provide external dependencies can adopt these rules to annotate external deps with package metadata, there is no canonical way of doing this for Bazel modules themselves. 

I would like to discuss and collect the steps we need to take to improve this situation. My ideas so far:

### Add a [purl](https://github.com/package-url/purl-spec) type for Bazel modules

This would provide a canonical way to refer to a given Bazel module, even if that module is not in the BCR. I prepared a draft of a specification of such a type [on this branch](https://github.com/package-url/purl-spec/compare/master...fmeum:purl-spec:patch-1).

### Make it easy for public Bazel modules to attach `package_info`

Compared to license information, which may easily vary on a per-target basis, package metadata as specified in `package_info` would usually apply to the entire module. By adding a small helper macro to `rules_license`, the required integration for Bazel modules could be reduced to "just":

```starlark
# MODULE.bazel
bazel_dep(name = "rules_license", version = "0.0.8")

# REPO.bazel
repo(default_package_metadata = ["//:package_info"])

# BUILD.bazel
load("@rules_license//rules:current_module_package_info.bzl", "current_module_package_info")

current_module_package_info(name = "package_info")
```

The `current_module_package_info` macro would rely on `native.module_name()` and `native.module_version()` to populate defaults for all the fields on `package_info` automatically (e.g., the purl).

Bazel registry tooling could be extended add the boilerplate automatically to new modules.

Modules with more complex requirements could still override the default package metadata on a per-package or per-target basis or choose to not use the macro.

**Comment by fmeum** [comment]
cc @Wyverald @meteorcloudy @aiuto @shs96c @keith @meisterT

**Reply by meteorcloudy** [reply]
Also cc @fweikert who is going to work on supply chain security for Bazel

Thanks for putting this together?
How is this purl specification used? Do you expect other toolings using this specification besides Bazel itself? Is it widely adopted in the OSS world?

**Reply by darkrift** [reply]
PURL is already part of cyclonedx and spdx specifications.

[Ref to cyclonedx 1.4 json schema](https://github.com/CycloneDX/specification/blob/e12ac5181bf3e0416660252e6da5ae80df673ce1/schema/bom-1.4.schema.json#L375-L380)
[SPDX spec ref](https://spdx.github.io/spdx-spec/v3.0/model/Software/Properties/packageUrl/)

**Reply by darkrift** [reply]
Support in rules_license was also added by https://github.com/bazelbuild/rules_license/pull/143

**Reply by fmeum** [reply]
Yes, purl is the closest to a universal format we will get. rules_license has (unreleased) support for it: https://github.com/bazelbuild/rules_license/commit/4f8f5f2d535fa78c6e9f91e4958b1a1b04fa22b1

**Reply by meteorcloudy** [reply]
Nice, thanks for the context!

**Reply by aiuto** [reply]
That just got merged. There is another PR in flight to use it in the sbom maker.

**Comment by shs96c** [comment]
@fmeum I'm not sure that your assertion about package info being correct for an entire module is true. For example, in a monorepo, there may be several artifacts that could be produced that others could depend on (eg. `java_export` targets), and each of those could have different package info. That is, package info could be per-target.

**Reply by fmeum** [reply]
It's not meant to be an assertion about all modules, just about the common case. I added a paragraph to that section to explain other options for modules with more complex requirements. I'm just trying to keep the amount of boilerplate for the common case as low as possible.

**Comment by sluongng** [comment]
I like the idea about `current_module_package_info`. 

On the REPO.bazel file: what would be the right way to set this up for repositories with 2 or more licenses? For example: https://gitlab.com/gitlab-org/gitlab uses a different license for `/ee` directory. Or https://github.com/open-telemetry/opentelemetry-go which contains multiple go modules with different versions underneath.

**Reply by fmeum** [reply]
> https://gitlab.com/gitlab-org/gitlab

It's not very likely that this repo will become a Bazel module in the BCR anytime soon, so this is more of an example of a complex end-user repo. As such, a `REPO.bazel` file at the top-level could set the OSS license and the individual `BUILD.bazel` files in the subdirectories of `/ee` could set another license via `package`. I would generally expect a repo of this size to use a Gazelle language or some other mechanism to generate `package_info` targets anyway

> https://github.com/open-telemetry/opentelemetry-go

This repo should be consumed as Go modules via the facilities provided by `gazelle`'s `go_deps` extension. Each Go submodule would have its own `pkg:go` purl. It wouldn't be related to what we do for Bazel BCR modules.

**Comment by shs96c** [comment]
@fmeum reading the suggested purl spec, it looks good. I would point out that [`repository_url` and `download_url` are already known qualifiers on a purl](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs) so the use of `namespace` may cause duplicate information to be included in the purl. That is, `pkg:bazel/example.org/bazel-registry/curl@8.8.0` may be rewritten as `pkg:bazel/curl@8.8.0?repository_url=example.org/bazel-registry`

I think there might be a valid counter argument that this is overloading `repository_url` and so the `namespace` is necessary, and that's entirely valid.

Within a single repo, it is also useful to know which target produced each artifact. It would be good to have a mechanism to (optionally) include that information too. Perhaps a `target` qualifier?

**Reply by fmeum** [reply]
> @fmeum reading the suggested purl spec, it looks good. I would point out that [`repository_url` and `download_url` are already known qualifiers on a purl](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs) so the use of `namespace` may cause duplicate information to be included in the purl. That is, `pkg:bazel/example.org/bazel-registry/curl@8.8.0` may be rewritten as `pkg:bazel/curl@8.8.0?repository_url=example.org/bazel-registry`

Thanks, I didn't know that `repository_url` was meant to refer to a package registry. It does look like the better way of encoding alternative registries. I will update the draft.

> Within a single repo, it is also useful to know which target produced each artifact. It would be good to have a mechanism to (optionally) include that information too. Perhaps a target qualifier?

I was already considering including a way to refer to a particular target. Now that you bring it up as well, let me just include it in the spec.

**Reply by fmeum** [reply]
I updated the spec to use `repository_url`. I also found `subpath` to work well for what `target` would have specified. PTAL.

**Reply by shs96c** [reply]
LGTM. Thank you for making the changes.

**Reply by fmeum** [reply]
I made some small edits and added test cases. I started a [draft PR](https://github.com/package-url/purl-spec/pull/317) and would submit it after discussing this topic at the next Rules Authors SIG meeting.

**Comment by aiuto** [comment]
Building `pacakge_info` dynamically sounds promising. The next step would be to tie it to build targets via  `package(default_package_metadata)`. 
How do you propose to do that?  Or, if that is not what you intend, how do you go about finding it doing an aspect traversal from a build target to the package data?

**Reply by fmeum** [reply]
That's this part:

```starlark
# REPO.bazel
repo(default_package_metadata = ["//:package_info"])
```

This sets a repo-wide default, which is more straightforward then adding a `package` call to all packages. Individual packages and targets can of course still override this.

**Reply by fmeum** [reply]
I sent a draft PR implementing this macro: https://github.com/bazelbuild/rules_license/pull/147

**Comment by aiuto** [comment]
Cool.  FWIW... I am currently very interested in making SBOM generation
work. But, of course, the user here is using WORKSPACE rather than bzlmod,
so I want to try your thing along with a workspace rule. Having both would
be great.

**Comment by fmeum** [comment]
@mzeren-vmw We discussed this in the Rules Authors SIG meeting, but would also be interested in your opinion on the purl spec addition.
